### ISSUE TITLE
Change Request: Top-level-await and strict-mode configuration

### ISSUE THREAD

**Issue by erosman** [eslint/eslint#19881]
### ESLint version

v9.29.0

### What problem do you want to solve?

Node.js, Greasemonkey, Tampermonkey and Violentmonkey (and FireMonkey 3.0+), wrap code in an IIFE before execution. I imagine `globalReturn` option was created for this purpose as otherwise the code would have had a `SyntaxError: return not in function`.

Top-level `await` is also becoming popular in userscripts which is achieved by adding `async` to the aforementioned IIFE.

Setting `sourceType: "module"` works for Top-level `await` but causes some issues since linter will evaluate the code in strict mode.

Possible solutions could be, adding a permanent and/or inline configs.

- allow Top-level `await` as an option to be used in `sourceType: 'script'`, similar to `globalReturn`
- Alternatively, support inline config such as: 
```js
/* eslint disable-top-level-await */
/* eslint enable-top-level-await */
/* eslint disable-strict-mode */
/* eslint enable-strict-mode */
```

Personally, I find the inline config approach more appealing.

See also:
- [Is it possible to have top-level await in `sourceType: "script"`?](https://github.com/eslint/eslint/discussions/19854)
- [@top-level-await](https://violentmonkey.github.io/api/metadata-block/#top-level-await)

### What do you think is the correct solution?

see above

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by fasttime** [comment]
It's not possible to allow parsing top-level `await` in script mode in Espree, although new versions of Acorn provide the option `allowAwaitOutsideFunction` for this purpose. To parse TLA in scripts with ESLint, you could probably write a [custom processor](https://eslint.org/docs/latest/extend/custom-processors) that wraps the code in an async IIFE.

**Comment by nzakas** [comment]
@fasttime please remember to move the issue into "Triaging" when you are triaging it.

**Comment by fasttime** [comment]
I think this is an interesting proposal. Besides the browser extensions mentioned in the issue, the [Node.js REPL](https://nodejs.org/en/learn/command-line/how-to-use-the-nodejs-repl) and the consoles of all browsers treat their input as script code wrapped in an async IIFE. This allows one to evaluate expressions such as `await ++"0"[0]` which are otherwise only valid in async functions in a script, in non-strict mode.

Enabling top-level await in scripts would mean updating Espree, `eslint-scope`, and significant portions of ESLint's core and built-in rules to make everything compatible with the new mode. Since ESLint doesn't support environments like browser extensions or interactive consoles, that's a substantial effort for very little immediate return. I appreciate the suggestion, but I'm not convinced it's the right time to commit to this change — it could slow down other priorities. I'd be interested to hear the TSC's perspective. @eslint/eslint-tsc

**Comment by erosman** [comment]
I would like to point out that the proposal has two aspects. Setting `sourceType: "module"` allows top-levet `await` but sets `"use strict"`

Possible approaches could be:
- allow top-level `await` in `sourceType: 'script'`
- Override strict mode in `sourceType: "module"`

I must also point out that there could be other (albeit minor) discrepancies e.g. `sourceType: "module"` would allow `import` while it wouldn’t be appropriate in the intended context.

### Alternative Approach

Does ESLint have a error-suppressing procedure?
Is there a process where ESLint would function as it should (without changing Espree) but errors can be suppressed/filtered at output stage?

I used to do that manually with CodeMirror and JSHint i.e. filter out unwanted errors before there are displayed.

**Comment by nzakas** [comment]
This would move us in the opposite direction of where we've been heading. We introduced `sourceType=commonjs` so we could avoid using `globalReturn` and eventually remove it. (See https://github.com/eslint/js/issues/525) We also had a similar discussion around top-level `using` in CommonJS: https://github.com/eslint/js/issues/662.

We had a lot of trouble previous with trying to mix modes (CommonJS vs. ESM) with features (strict mode, top-level `await`).

So, I'm not in favor of adding more fine-grained controls to pick and choose which features are enabled. I feel like this is too much added complexity to address what is an edge case. Should this edge case move to become a more common case (for instance, if CommonJS decides to support top-level await) then we can always revisit.

**Comment by fasttime** [comment]
> ### Alternative Approach
> 
> Does ESLint have a error-suppressing procedure? Is there a process where ESLint would function as it should (without changing Espree) but errors can be suppressed/filtered at output stage?
> 
> I used to do that manually with CodeMirror and JSHint i.e. filter out unwanted errors before there are displayed.

I think I do not understand how suppressing errors (what errors?) would be an alternative to the what the current issue is proposing. If you are looking for a solution to a specific problem, I'd suggest opening a [discussion](https://github.com/eslint/eslint/discussions) or asking in [Discord](https://eslint.org/chat).

As for the rest, I agree with @nzakas that this feature seems too complex for the utility it brings. I'd be happy to reconsider this in the future, should top-level await in scripts become more common. Thanks!
